### PR TITLE
feat: support specify mac addr via annotation

### LINKED ISSUE #190: support MAC assignment through pod annotation
Issue Description
---

Type: *feature request*

### Describe what feature you want
1. assign MAC address in pod annotation
2. check MAC conflict

### PR DESCRIPTION
**Pull request by sjtufl**
Pull Request Description
---

### Describe what this PR does / why we need it

Introduce a new feature: support specifying MAC address via pod annotation.

### Does this pull request fix one issue?

Fixes #190 

### Describe how you did it

1. Verify the validness of specified MAC address in validating webhook;
2. Check whether the specified MAC address is in conflict with existing IP instances.

### Describe how to verify it
Specify a static MAC address via annotation `"networking.alibaba.com/specified-mac"` on Pod. 

### Special notes for reviews

### CODE REVIEW

**Comment by hhyasdf** [comment]
What's the new feature for? I'm asking this because pod MAC address is not "exported" for hybridnet, in another word, a pod will always use the MAC address of its local node to communicate with the pods on other nodes.

**Comment by sjtufl** [comment]
> What's the new feature for? I'm asking this because pod MAC address is not "exported" for hybridnet, in another word, a pod will always use the MAC address of its local node to communicate with the pods on other nodes. The same as CNI ipvlan.

Yes, it's true that user-specified MAC address is not necessary in terms of communicating with pods on other nodes. But we've also seen the need in some "non-communication" scenarios, e.g., internal auditing or generating product license. So I think it's more like a nice-to-have feature with no interference in core data-plane functionality.

**Comment by hhyasdf** [comment]
Could you please provide a sample usage for this feature in "Describe how to verify it" section? 

I've noticed that a pod's MAC address can be specified with an `"networking.alibaba.com/specified-mac"` annotation. What will happen if the pod is create as a part of deployment? One specified MAC address for a pod seems not work out. 

Maybe it should be implemented similarly as `"networking.alibaba.com/ip-pool"`, which we already have.

File: pkg/controllers/networking/pod_controller.go [hunk 1]
```diff
@@ -123,6 +124,25 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result
 		return ctrl.Result{}, nil
 	}
 
+	// It is not allowed that specified mac address is in conflict with existing IPInstances.
```
Comment (mars1024) [inline pkg/controllers/networking/pod_controller.go:127]: can you check this in `createIP` of `IPAMStore`?

File: pkg/controllers/networking/pod_controller.go [hunk 2]
```diff
@@ -536,6 +556,17 @@ func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) (err error) {
 		return err
 	}
 
+	// Init mac indexer for IPInstances.
+	if err = mgr.GetFieldIndexer().IndexField(context.TODO(), &networkingv1.IPInstance{}, indexerFieldMac, func(obj client.Object) []string {
```
Comment (mars1024) [inline pkg/controllers/networking/pod_controller.go:560]: After the latest commit, this indexer registration may be moved to another place.

**Comment by sjtufl** [comment]
> Could you please provide a sample usage for this feature in "Describe how to verify it" section?

OK, edited!

> I've noticed that a pod's MAC address can be specified with an `"networking.alibaba.com/specified-mac"` annotation. What will happen if the pod is create as a part of Deployment or StatefulSet? One specified MAC address for a pod seems not work out.
> 
> Maybe it should be implemented similarly as the annotation of `"networking.alibaba.com/ip-pool"`, which we already have in [wiki docs](https://github.com/alibaba/hybridnet/wiki/Static-pod-ip-addresses-for-StatefulSet).

Current implementation is only for single pod scenario and does not support workloads like Deployment or StatefulSet. Unlike IP retain, specifying static MAC address shouldn't be limited to stateful workloads. I think that specifying a "MAC-Pool" would be great, and it will require some additional work.

File: pkg/controllers/networking/pod_controller.go (hunk 1 above)
Comment (sjtufl) [reply pkg/controllers/networking/pod_controller.go]: Okay! It makes conflicts checked in a "global" manner.

**Comment by hhyasdf** [comment]
> > Could you please provide a sample usage for this feature in "Describe how to verify it" section?
> 
> OK, edited!
> 
> > I've noticed that a pod's MAC address can be specified with an `"networking.alibaba.com/specified-mac"` annotation. What will happen if the pod is create as a part of Deployment or StatefulSet? One specified MAC address for a pod seems not work out.
> > Maybe it should be implemented similarly as the annotation of `"networking.alibaba.com/ip-pool"`, which we already have in [wiki docs](https://github.com/alibaba/hybridnet/wiki/Static-pod-ip-addresses-for-StatefulSet).
> 
> Current implementation is only for single pod scenario and does not support workloads like Deployment or StatefulSet. Unlike IP retain, specifying static MAC address shouldn't be limited to stateful workloads. I think that specifying a "MAC-Pool" would be great, and it will require some additional work.

Fine. If only for a single pod, maybe this should be considered as a feature for developers (of k8s controllers) but not for all the users. Looks good for me, cc @mars1024 .

**Comment by hhyasdf** [comment]
Please rebase `main` branch to fix the golangci-lint error.

**Comment by sjtufl** [comment]
This PR is closed since IPAM module is refactored a lot and leads to multiple conflicts. I will open another PR on supporting mac address pool assignment.

**Pull request closed without merging**